File: packages/python/minions/storage/memory_storage_adapter.py
"""
minions.storage.memory_storage_adapter
=======================================
In-memory storage adapter — useful for testing and ephemeral workloads.
"""

from __future__ import annotations

from typing import Optional

from ..types import Minion
from .adapter import StorageAdapter, StorageFilter


def _apply_filter(minions: list[Minion], filter: StorageFilter) -> list[Minion]:
    """Apply a :class:`StorageFilter` to a list of minions."""
    result = minions

    if not filter.include_deleted:
        result = [m for m in result if not m.deleted_at]
    if filter.minion_type_id is not None:
        result = [m for m in result if m.minion_type_id == filter.minion_type_id]
    if filter.status is not None:
        result = [m for m in result if m.status == filter.status]
    if filter.tags:
        result = [m for m in result if all(t in (m.tags or []) for t in filter.tags)]

    result = result[filter.offset:]

    if filter.limit is not None:
        result = result[: filter.limit]

    return result


class MemoryStorageAdapter(StorageAdapter):
    """
    Simple in-memory storage adapter.

    All data is stored in a ``dict`` and is lost when the process exits.
    This is the default adapter when no persistence is required and is well
    suited for unit tests.
    """

    def __init__(self) -> None:
        self._store: dict[str, Minion] = {}

    async def get(self, id: str) -> Optional[Minion]:
        return self._store.get(id)

    async def set(self, minion: Minion) -> None:
        self._store[minion.id] = minion

    async def delete(self, id: str) -> None:
        self._store.pop(id, None)

    async def list(self, filter: Optional[StorageFilter] = None) -> list[Minion]:
        all_minions = list(self._store.values())
        if filter is None:
            return [m for m in all_minions if not m.deleted_at]
        return _apply_filter(all_minions, filter)

    async def search(self, query: str) -> list[Minion]:
        if not query.strip():
            return await self.list()

        tokens = query.lower().split()
        all_minions = [m for m in self._store.values() if not m.deleted_at]

        return [
            m for m in all_minions
            if all(token in (m.searchable_text or m.title.lower()) for token in tokens)
        ]

File: packages/core/src/storage/MemoryStorageAdapter.ts
/**
 * @module minions-sdk/storage/memory
 * In-memory storage adapter — useful for testing and ephemeral workloads.
 */

import type { Minion } from '../types/index.js';
import type { StorageAdapter, StorageFilter } from './StorageAdapter.js';
import { applyFilter } from './filterUtils.js';

/**
 * Simple in-memory storage adapter.
 *
 * All data is stored in a `Map` and is lost when the process exits.
 * This is the default adapter when no persistence is required and is well
 * suited for unit tests.
 *
 * Supports the optional attachment file operations using an in-memory store.
 */
export class MemoryStorageAdapter implements StorageAdapter {
  private store: Map<string, Minion> = new Map();
  private files: Map<string, Map<string, Uint8Array>> = new Map();

  async get(id: string): Promise<Minion | undefined> {
    return this.store.get(id);
  }

  async set(minion: Minion): Promise<void> {
    this.store.set(minion.id, minion);
  }

  async delete(id: string): Promise<void> {
    this.store.delete(id);
    this.files.delete(id);
  }

  async list(filter?: StorageFilter): Promise<Minion[]> {
    const all = Array.from(this.store.values());
    if (!filter) {
      return all.filter((m) => !m.deletedAt);
    }
    return applyFilter(all, filter);
  }

  async search(query: string): Promise<Minion[]> {
    if (!query.trim()) return this.list();

    const tokens = query.toLowerCase().split(/\s+/).filter(Boolean);
    const all = Array.from(this.store.values()).filter((m) => !m.deletedAt);

    return all.filter((m) => {
      const text = (m.searchableText ?? m.title).toLowerCase();
      return tokens.every((token) => text.includes(token));
    });
  }

  // ── Attachment file operations ─────────────────────────────────────────

  async putFile(id: string, filename: string, data: Uint8Array): Promise<void> {
    let bucket = this.files.get(id);
    if (!bucket) {
      bucket = new Map();
      this.files.set(id, bucket);
    }
    bucket.set(filename, new Uint8Array(data));
  }

  async getFile(id: string, filename: string): Promise<Uint8Array | undefined> {
    return this.files.get(id)?.get(filename);
  }

  async deleteFile(id: string, filename: string): Promise<void> {
    this.files.get(id)?.delete(filename);
  }

  async listFiles(id: string): Promise<string[]> {
    const bucket = this.files.get(id);
    return bucket ? Array.from(bucket.keys()) : [];
  }
}
